hail/src/main/scala/is/hail/expr/Parser.scala

```diff
@@ -505,18 +493,19 @@ object Parser extends JavaTokenParsers {
       AggSignature(op, ctorArgTypes.map(t => -t), initOpArgTypes.map(_.map(t => -t)), seqOpArgTypes.map(t => -t))
     }
 
-  def ir_named_value_exprs: Parser[Seq[(String, ir.IR)]] = rep(ir_named_value_expr)
+  def ir_named_value_exprs(env: IRParserEnvironment): Parser[Seq[(String, ir.IR)]] =
+    rep(ir_named_value_expr(env))
 
-  def ir_named_value_expr: Parser[(String, ir.IR)] =
-    "(" ~> ir_identifier ~ ir_value_expr <~ ")" ^^ { case n ~ x => (n, x) }
+  def ir_named_value_expr(env: IRParserEnvironment): Parser[(String, ir.IR)] =
+    "(" ~> ir_identifier ~ ir_value_expr(env) <~ ")" ^^ { case n ~ x => (n, x) }
 
   def ir_opt[T](p: Parser[T]): Parser[Option[T]] = p ^^ {
     Some(_)
   } | "None" ^^ { _ => None }
 
-  def ir_value_expr: Parser[ir.IR] = "(" ~> ir_value_expr_1 <~ ")"
+  def ir_value_expr(env: IRParserEnvironment): Parser[ir.IR] = "(" ~> ir_value_expr_1(env) <~ ")"
```

Comment:
this should use IRParserEnvironment.theEnv, not parameterize the env. I benchmarked this stuff a couple months ago and it was *disastrous* when written this way, so moved to the static object reference.

Reply:
this change is awesome though and I fully support it

Reply:
I think this is very confusing with global state (if it is even possible).  I'd prefer to just write this directly without Scala's parser combinator magic, than use global state.

Reply:
What was your benchmark?

Reply:
I generated huge things and parsed them and saw that take ~10s, then made the global state change and brought it back down to ~100ms

Reply:
10s.  Here's what I suggest:
 - let this go in, then Jackie can start pushing the Matrix/Table functions into the IR rather than calling into Java, and
 - in parallel write a parser (directly, no combinators) which won't allocate at all.
in which case we'd have to live with this for few days.  OK?

Reply:
fine with me - I made that comment before I read the full change. This is definitely worth the slowdown again.

Reply:
OK, and I just walked through turning this into a recursive descent parser with @jigold, so that will fix this problem once and for all.